vllm/worker/enc_dec_model_runner.py

```diff
@@ -0,0 +1,478 @@
+        * `encoder_seq_lens`
+        * `encoder_seq_lens_tensor`
+        * `max_encoder_seq_len`
+        * `cross_slot_mapping`
+        * `cross_block_tables`
+
+        Constructs a new model inputs data structure, based on
+        (1) the existing fields in the `model_inputs` argument,
+        and (2) the following addition fields that are
+        computed (or in the case of `attn_metadata`, updated) 
+        by this function:
+        * attn_metadata
+        * encoder_input_tokens
+        * encoder_input_positions
+
+        Arguments:
+
+        * seq_group_metadata_list: list of sequence groups for which to
+                                   compute inputs
+        * model_inputs: model inputs data structure with decoder-oriented
+                        fields already computed.
+
+        Return:
+
+        * Updated model inputs data structure
+        """
+
+        encoder_input_tokens: List[int] = []
+        encoder_input_positions: List[int] = []
+        encoder_seq_lens: List[int] = []
+        context_lens: List[int] = []
+        query_lens: List[int] = []
+        cross_slot_mapping: List[int] = []
+        cross_block_tables: List[List[int]] = []
+        prompt_adapter_index_mapping: List[int] = []
+        prompt_adapter_prompt_mapping: List[int] = []
+        prompt_adapter_requests: Set[PromptAdapterRequest] = set()
+
+        if len(seq_group_metadata_list) == 0:
+            # Leave the encoder/cross-attention input
+            # fields at default values if the seq group
+            # metadata list arg is an empty list
+            return model_input
+
+        for seq_group_metadata in seq_group_metadata_list:
+            is_prompt = seq_group_metadata.is_prompt
+
+            encoder_seq_data = seq_group_metadata.encoder_seq_data
+            cross_block_table = seq_group_metadata.cross_block_table
+            context_len = (0 if is_prompt else encoder_seq_data.get_len())
+
+            seq_len = encoder_seq_data.get_len()
+
+            block_table = []  # Default for memory profiling
+            if is_prompt:
+                tokens = encoder_seq_data.get_token_ids()[context_len:seq_len]
+            else:
+                # Optimization. get_token_ids requires the entire copy of
+                # tokens.
+                tokens = [encoder_seq_data.get_last_token_id()]
+
+                if cross_block_table is not None:
+                    # Decode
+                    block_table = cross_block_table
+
+            block_table = (
+                []  # Memory profiling
+                if
+                (is_prompt or cross_block_table is None) else cross_block_table
+            )  # Decode
+
+            cross_block_tables.append(block_table)
+
+            encoder_seq_lens.append(seq_len)
+            context_lens.append(context_len)
+            query_len = seq_len - context_len
+            query_lens.append(query_len)
+            encoder_input_tokens.extend(tokens)
```

Comment:
I am wondering if encoder_input_tokens should be the same for both is_prompt and not is_prompt case. The encoder_input_tokens does not change whether its prefill or decode. Hence wondering if encoder_input_tokens should be same for both case. Currently you seem to treat them differently (L335 and L339). Same comment for  encoder_input_positions

Reply:
Understand where your coming from. As you saw, the current behavior of the code is that the encoder tokens sequence is empty in the decode phase. The rationale for this is to maintain consistency with decoder-ony behavior: in a decoder-only model, during the decode phase, the input token sequence produced by model runner is the concatenation of decoded tokens for each sequence from the last step. i.e. for 10 sequences, in which each one had a single decoded token in the last iteration, then model runner will construct a model input token sequence with 10 tokens.

In the encoder/decoder scenario, this behavior is exactly the same with regards to how the decoder input sequence is constructed. However, what about the encoder input sequence? Well, the encoder input sequence is only input to the encoder during prefill, and the encoder hidden states are static (read-only) after prefill; no new encoder tokens are ever decoded. Therefore, it makes sense that in each decode-phase iteration, the encoder "input tokens" list is of length zero, because no new encoder input tokens were decoded in the previous iteration.